doc/neps/nep-0033-array-creation-dispatch-with-array-function.rst

```diff
@@ -0,0 +1,186 @@
+is twofold:
+
+1. The object will have no use in the downstream library's implementation; and
+2. Simplifies adoption of array creation by those libraries already opting-in
+   to implement the ``__array_function__`` protocol, thus removing the
+   requirement to explicitly opt-in for all array creation functions.
+
+Downstream libraries thus shall _NOT_ include the ``like=`` keyword to their
+array creation APIs, which is a NumPy-exclusive keyword.
+
+Function Dispatching
+~~~~~~~~~~~~~~~~~~~~
+
+There are two different cases to dispatch: Python functions, and C functions.
+To permit ``__array_function__`` dispatching, one possible implementation is to
+decorate Python functions with ``overrides.array_function_dispatch``, but C
+functions have a different requirement, which we shall describe shortly.
+
+The example below shows a suggestion on how the ``asarray`` could be decorated
+with ``overrides.array_function_dispatch``:
+
+.. code:: python
+
+    def _asarray_decorator(a, dtype=None, order=None, like=None):
+        return (like,)
+
+    @set_module('numpy')
+    @array_function_dispatch(_asarray_decorator)
+    def asarray(a, dtype=None, order=None, like=None):
+        return array(a, dtype, copy=False, order=order)
+
+Note in the example above that the implementation remains unchanged, the only
+difference is the decoration, which uses the new ``_asarray_decorator`` function
+to instruct the ``__array_function__`` protocol to dispatch if ``like`` is not
+``None``.
+
+We will now look at a C function example, and since ``asarray`` is anyway a
+specialization of ``array``, we will use the latter as an example now. As
+``array`` is a C function, currently all NumPy does regarding its Python source
+is to import the function and adjust its ``__module__`` to ``numpy``. The
+function will now be decorated with a specialization of
+``overrides.array_function_from_dispatcher``, which shall take care of adjusting
+the module too.
+
+.. code:: python
+
+    array_function_nodocs_from_c_func_and_dispatcher = functools.partial(
+        overrides.array_function_from_dispatcher,
+        module='numpy', docs_from_dispatcher=False, verify=False)
+
+    @array_function_nodocs_from_c_func_and_dispatcher(_multiarray_umath.array)
+    def array(a, dtype=None, copy=True, order='K', subok=False, ndmin=0, like=None):
+        return (like,)
+
+There are two downsides to the implementation above for C functions:
+
+1.  It creates another Python function call; and
+2.  To follow current implementation standards, documentation should be attached
+    directly to the Python source code.
+
+Alternatively for C functions, the implementation of ``like=`` could be moved
+into the C implementation itself. This is not the primary suggestion here due
+to its inherent complexity which would be difficult too long to describe in its
+entirety here, and too tedious for the reader. However, we leave that as an
+option open for discussion.
```

Comment:
I think we may want to go the C route (or get out the cython idea again). (The main issue in python is that there is no simple fast kwarg parsing solution available right now; python has the clinic internally, but that does not really help us).

The thing is that if this is in python, it will add at least one more function overhead call, which is fairly significant for a no-op. For many library functions it will not matter, but for the simple numpy functions it may.

Reply:
I suspected people would lean this way, that's why I suggested this, from my perspective, I'm fine with either way. Maybe would be good to wait and see what others prefer too?

If we go the C path, IMO it doesn't make sense to extend the NEP quite significantly to point out all the adaptations needed to allow this. Unless people disagree with my point, I  would only move the suggestion for Python implementation to a "rejected implementation proposal" section and mention that we are going the C way, but not describe all its details, as those details will likely not aggregate any important information to non-NumPy developers using that as a reference.

Reply:
The only part that is currently not implemented in C is the `public_api` helper function from inside`numpy.core.overrides.array_function_dispatch`:
```python
        def public_api(*args, **kwargs):
            relevant_args = dispatcher(*args, **kwargs)
            return implement_array_function(
                implementation, public_api, relevant_args, args, kwargs)
```

Maybe we could save the overhead of parsing `args` and `kwargs` in `public_api` if we wrote it in C? Honestly I'm not entirely sure.

This uses a closure over `implementation` and `public_api`, which is probably the main reason why it wasn't written in C in the first place.

Reply:
Sorry for being unresponsive. I have been thinking about this, what I suppose could work is to add the `like` kwarg to the C implementation of each function, and then do the inverse process:

1. If `like` is valid, call the C implementation of override;
1. The C implementation of override would have a callback to the Python implementation which would dispatch to the downstream library.

Admittedly, I don't know much about C Python and the current C implementation of the `__array_function__` override, so I don't know whether this is something that is possible to achieve. Besides this, I also don't see any other alternatives to doing it the Python way, as per my proposal in the NEP itself. Thoughts on this @shoyer ?